src/sentry/workflow_engine/models/detector.py

```diff
@@ -111,3 +126,80 @@ def __init__(self, detector: Detector):
     @abc.abstractmethod
     def evaluate(self, data_packet: DataPacket[T]) -> list[DetectorEvaluationResult]:
         pass
+
+
+class StatefulDetectorHandler(DetectorHandler[T], abc.ABC):
+    def build_dedupe_value_key(self, group_key: str | None) -> str:
+        if group_key is None:
+            group_key = ""
+        return f"{self.detector.id}:{group_key}:dedupe_value"
+
+    def build_counter_value_key(self, group_key: str | None, counter_name: str) -> str:
+        if group_key is None:
+            group_key = ""
+        return f"{self.detector.id}:{group_key}:{counter_name}"
+
+    def bulk_get_detector_state(
+        self, state_updates: list[DetectorStateData]
+    ) -> dict[str | None, DetectorState]:
+        group_keys = {update.group_key for update in state_updates}
+        query_filter = Q(
+            detector_group_key__in=[group_key for group_key in group_keys if group_key is not None]
+        )
+        if None in group_keys:
+            query_filter |= Q(detector_group_key__isnull=True)
+
+        return {
+            detector_state.detector_group_key: detector_state
+            for detector_state in self.detector.detectorstate_set.filter(query_filter)
+        }
+
+    def commit_state_update_data(self, state_updates: list[DetectorStateData]):
+        detector_state_lookup = self.bulk_get_detector_state(state_updates)
+        created_detector_states = []
+        updated_detector_states = []
+        dedupe_values = []
+        group_counter_updates = {}
+        for state_update in state_updates:
+            detector_state = detector_state_lookup.get(state_update.group_key)
+            if not detector_state:
+                created_detector_states.append(
+                    DetectorState(
+                        detector_group_key=state_update.group_key,
+                        detector=self.detector,
+                        active=state_update.active,
+                        state=state_update.status,
+                    )
+                )
+            elif (
+                state_update.active != detector_state.active
+                or state_update.status != detector_state.state
+            ):
+                detector_state.active = state_update.active
```

Comment:
should this just be `detector_state = state_update` incase there is new data in other fields?

Reply:
We have to copy the data here from the update state to the model.

It's possible that this abstraction is a little weird now that we've changed all the DIWA stuff. I'll revisit it in later prs